File: Data Structures & Algorithms/sort-colors/submission-4.cpp
class Solution {
public:
    void sortColors(vector<int>& nums) {
        vector<int> count(2, 0);

        for (int num:nums) {
            if (num == 2) continue;
            count[num]++;
        }

        for (int i = 0; i<nums.size(); i++) {
            if (i < count[0]) {
                nums[i] = 0;
            } else if (i < count[1]+count[0]) {
                nums[i] = 1;
            } else {
                nums[i] = 2;
            }
        }
    }
};

File: Data Structures & Algorithms/longest-increasing-subsequence/submission-4.py
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        memo = [[-1]*n for _ in range(n)]

        def dfs(i, j):
            if i == n:
                return 0
            if memo[i][j] != -1:
                return memo[i][j]
            
            LIS = dfs(i+1, j)

            if j == -1 or nums[i] > nums[j]:
                LIS = max(LIS, 1 + dfs(i+1, i))

            memo[i][j] = LIS

            return LIS

        return dfs(0, -1)